apps/req.c

```diff
@@ -1751,15 +1751,19 @@ int do_X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md,
 #endif
 
     rv = do_sign_init(mctx, pkey, md, sigopts);
-    if (rv > 0)
+    if (rv > 0) {
         rv = X509_sign_ctx(x, mctx);
 #ifndef OPENSSL_NO_SM2
-    /* only in SM2 case we need to free the pctx explicitly */
-    if (ec_pkey_is_sm2(pkey)) {
-        pctx = EVP_MD_CTX_pkey_ctx(mctx);
-        EVP_PKEY_CTX_free(pctx);
-    }
+        /*
+         * only in SM2 case we need to free the pctx explicitly
+         * if do_sign_init() fails, pctx is already freed in it
```

Comment:
Unfortunately the `X509_sign_ctx` requires to use the `pctx` assigned in the `mctx`, so it can only be freed after the `X509_sign_ctx` call.

Reply:
Hmm. Perhaps then do_sign_init should pass back the pctx (or NULL if it didn't create its own one). Then you can just free it unconditionally

Reply:
Yes, that was in fact considered when the code was firstly created. But I found that would make the code even harder to understand. Since there is an `EVP_MD_CTX_pkey_ctx` API we could utilize, I think we should use that one.

Reply:
So...considerations?

Reply:
Ok - if you have considered it then no problem.